// @builder.io/mitosis-cli/dist/build/helpers/inputs-extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INPUT_EXTENSION_IMPORT_REGEX = exports.INPUT_EXTENSION_REGEX = exports.checkIsMitosisComponentFilePath = exports.INPUT_EXTENSIONS_ARRAY = exports.INPUT_EXTENSIONS = void 0;
var Array_1 = require("fp-ts/lib/Array");
exports.INPUT_EXTENSIONS = {
    jsx: ['.lite.tsx', '.lite.jsx'],
    svelte: ['.svelte'],
};
exports.INPUT_EXTENSIONS_ARRAY = (0, Array_1.flatten)(Object.values(exports.INPUT_EXTENSIONS));
// check if filePath ends with one of the values of INPUT_EXTENSIONS_ARRAY
var checkIsMitosisComponentFilePath = function (filePath) {
    return exports.INPUT_EXTENSIONS_ARRAY.some(function (extension) { return filePath.endsWith(extension); });
};
exports.checkIsMitosisComponentFilePath = checkIsMitosisComponentFilePath;
/**
 * Matches `.svelte`, `.lite.tsx`, `.lite.jsx` files (with optional `.jsx`/`.tsx` extension)
 */
exports.INPUT_EXTENSION_REGEX = /\.(svelte|(lite(\.tsx|\.jsx)?))/g;
// Adds trailing quotes to the end of import paths
exports.INPUT_EXTENSION_IMPORT_REGEX = /\.(svelte|(lite(\.tsx|\.jsx)?))(?<quote>['"])/g;
